consensus/reactor.go

```diff
@@ -487,6 +490,18 @@ func (conR *Reactor) sendNewRoundStepMessage(peer p2p.Peer) {
 	peer.Send(StateChannel, MustEncode(nrsMsg))
 }
 
+func (conR *Reactor) updateRoundStateRoutine() {
```

Comment:
put this in a timer, otherwise it seems like this could get hot? or what we really need is a condvar to notify us when it updates so we don't have to spin?

Reply:
If there's only one updater, you can also use a channel as a cheap condition variable.

Reply:
Yeah, it can although it's definitely less hot than it was before, since it previously was spinning once for every peer in a similar fashion.

Reply:
If we don't have a good way to signal updates, I think @tychoish is probably right that a timer would help. Goroutines that don't yield can starve the scheduler, and even a short timer (`100*time.Microsecond`) is enough to leave space.

Reply:
Fair enough. I feel sort of sad that we don't have a better way to signal updates, but in the absence of one, this is probably the best we can do.

Reply:
How do we know when we have an update? Is it just waiting for someone to set the round state? If so, we could use something like:

```go
var state struct {
   sync.Mutex
   done chan struct{}
   v *State
}

func update(st *State) {
   state.Lock()
   defer state.Unlock()
   state.v = st
   close(state.done)
   state.done = make(chan struct{})
}

func waitForUpdate(ctx context.Context) *State {
   state.Lock()
   ch := state.done
   state.Unlock()
   for {
       select {
       case <-ctx.Done():
          return
       case <-done:
          state.Lock()
          defer state.Unlock()
          ch = state.done
          return state.v
       }
   }
}
```

this is schematic of course, but works pretty well.

Reply:
@creachadair I'm almost certain that your idea is the right one. My concern is that at the moment, the code in `state.go` doesn't provide a signal that there's been an update at all of the times that we may want that signal. There is this [newStep](https://github.com/tendermint/tendermint/blob/0e130f5e2c096f487b79e82dbcb8ad2a3ac27aed/consensus/state.go) method on the State object that fires an event at almost all of the times we want want one, except, in the one that I'm worried is crucial to the osmosis issue: right after a proposed block has been completed. This code can be found here: https://github.com/tendermint/tendermint/blob/0e130f5e2c096f487b79e82dbcb8ad2a3ac27aed/consensus/state.go#L1914

To summarize the block completion code:
That code adds the block part to the the set of parts it has gathered from other validators. Then, if this code sees that it just received the completed block and has already received the +2/3 precommits for it, it immediately moves to execute the block against the app. All of this is done while holding the lock. 

Once the execution is completed, the `newStep` method is called. Therefore, using the signal generated in `newStep` would fail to send the full part set to the gossip code until after the block is executed. This would allow us to alleviate lock contention in many cases except not in the case of receiving and executing a full block. 

We could definitely instrument the `state.go` file, and I think that may be the best longterm bet, but my worry was really about getting that really right in just a fixup. That file is messy and I'm not sure the relationship between completing a block and gossiping it is well tested. Since it's so critical but also so hairy, I feel that getting something inelegant yet working may be the right thing to do for now. Open to being convinced otherwise.

Reply:
Ah, I see—so we currently rely on being able to "keep" the lock when we find a change?

I'm fine with maintaining the status quo, I was mainly just looking for ways to get the more general thing we want. I think it might be worthwhile to do the spin-throttling part, just to avoid pinning a CPU on that thread for periods when the round state doesn't get updated for a while, but as you say, there are a lot of dependencies.

Reply:
> Ah, I see—so we currently rely on being able to "keep" the lock when we find a change?

Exactly. I'd imagine that implicit assumption of this code was that no single 'step' would really take that much time so "keeping" the lock during both a data add + step was thought to generally be quite safe.

Reply:
That makes sense. If it's really just updating some state in response to the transition, that might not be a bad assumption.

Longer-term, we could get that effect by giving the "wait" step a callback:

```go
func waitAndUpdate(ctx context.Context, f func(ns *State)) {
   // ... as above, except call f under the lock instead of returning
}
```

Reply:
> That makes sense. If it's really just updating some state in response to the transition, that might not be a bad assumption.

Yeah, the assumption does cause issue when the `BeginBlock` takes a long time. Consensus holds the lock for the entirety of the date add+ the long `BeginBlock`.